src/decoder/mod.rs

```diff
@@ -188,217 +576,255 @@ impl<R: Read + Seek> DecoderImpl<R> {
-        let data = match mp3::Mp3Decoder::new(data) {
-            Err(data) => data,
-            Ok(decoder) => {
-                return Ok(Decoder(DecoderImpl::Mp3(decoder)));
-            }
-        };
+/// Converts a `File` into a `Decoder` with automatic optimizations.
+///
+/// This implementation:
+/// - Wraps the file in a `BufReader` for better performance
+/// - Sets `byte_len` from file metadata when available, which can improve seeking operations
+///
+/// # Examples
+/// ```no_run
+/// use std::fs::File;
+/// use std::convert::TryFrom;
+/// use rodio::Decoder;
+///
+/// let file = File::open("audio.mp3").unwrap();
+/// let decoder = Decoder::try_from(file).unwrap();
+/// ```
+impl TryFrom<std::fs::File> for Decoder<BufReader<std::fs::File>> {
+    type Error = DecoderError;
 
-        #[cfg(feature = "symphonia")]
-        {
-            let mss = MediaSourceStream::new(
-                Box::new(ReadSeekSource::new(data)) as Box<dyn MediaSource>,
-                Default::default(),
-            );
+    fn try_from(file: std::fs::File) -> Result<Self, Self::Error> {
+        let mut builder = DecoderBuilder::new();
+        if let Some(len) = file.metadata().ok().map(|m| m.len()) {
+            builder = builder.with_byte_len(len);
+        }
 
-            match symphonia::SymphoniaDecoder::new(mss, None) {
-                Err(e) => Err(e),
-                Ok(decoder) => Ok(Decoder(DecoderImpl::Symphonia(decoder))),
-            }
+        match builder.with_data(BufReader::new(file)).build()? {
+            DecoderOutput::Normal(decoder) => Ok(decoder),
+            DecoderOutput::Looped(_) => unreachable!("Builder defaults to non-looped"),
         }
-        #[cfg(not(feature = "symphonia"))]
-        Err(DecoderError::UnrecognizedFormat)
     }
+}
 
-    /// Builds a new looped decoder.
+impl<R: Read + Seek + Send + Sync + 'static> Decoder<R> {
+    /// Builds a new decoder with default settings.
     ///
     /// Attempts to automatically detect the format of the source of data.
-    pub fn new_looped(data: R) -> Result<LoopedDecoder<R>, DecoderError> {
-        Self::new(data).map(LoopedDecoder::new)
-    }
-
-    /// Builds a new decoder from wav data.
-    #[cfg(all(feature = "wav", not(feature = "symphonia-wav")))]
-    pub fn new_wav(data: R) -> Result<Decoder<R>, DecoderError> {
-        match wav::WavDecoder::new(data) {
-            Err(_) => Err(DecoderError::UnrecognizedFormat),
-            Ok(decoder) => Ok(Decoder(DecoderImpl::Wav(decoder))),
+    ///
+    /// # Errors
+    ///
+    /// Returns `DecoderError::UnrecognizedFormat` if the audio format could not be determined
+    /// or is not supported.
+    pub fn new(data: R) -> Result<Self, DecoderError> {
+        match DecoderBuilder::new().with_data(data).build()? {
+            DecoderOutput::Normal(decoder) => Ok(decoder),
+            DecoderOutput::Looped(_) => unreachable!("Builder defaults to non-looped"),
```

Comment:
splitting into `build` and `build_looped` would also take care of this.

Reply:
Done.